src/backend/fastapi_app/rag_advanced.py

```diff
@@ -65,44 +58,69 @@ async def run(
             tool_choice="auto",
         )
 
-        query_text, filters = extract_search_arguments(original_user_query, chat_completion)
+        query_text, filters = extract_search_arguments(chat_params.original_user_query, chat_completion)
+
+        return query_messages, query_text, filters
 
+    async def retreive_and_build_context(
+        self, chat_params: ChatParams, query_text: str | Any | None, filters: list
+    ) -> tuple[list[ChatCompletionMessageParam], list[Item]]:
+        """Retrieve relevant items from the database and build a context for the chat model."""
         # Retrieve relevant items from the database with the GPT optimized query
         results = await self.searcher.search_and_embed(
             query_text,
-            top=top,
-            enable_vector_search=vector_search,
-            enable_text_search=text_search,
+            top=chat_params.top,
+            enable_vector_search=chat_params.enable_vector_search,
+            enable_text_search=chat_params.enable_text_search,
             filters=filters,
         )
 
         sources_content = [f"[{(item.id)}]:{item.to_str_for_rag()}\n\n" for item in results]
         content = "\n".join(sources_content)
 
         # Generate a contextual and content specific answer using the search results and chat history
-        response_token_limit = 1024
         contextual_messages: list[ChatCompletionMessageParam] = build_messages(
             model=self.chat_model,
-            system_prompt=overrides.get("prompt_template") or self.answer_prompt_template,
-            new_user_content=original_user_query + "\n\nSources:\n" + content,
-            past_messages=past_messages,
-            max_tokens=self.chat_token_limit - response_token_limit,
+            system_prompt=chat_params.prompt_template,
+            new_user_content=chat_params.original_user_query + "\n\nSources:\n" + content,
+            past_messages=chat_params.past_messages,
+            max_tokens=self.chat_token_limit - chat_params.response_token_limit,
             fallback_to_default=True,
         )
+        return contextual_messages, results
+
+    async def run(
+        self,
+        messages: list[ChatCompletionMessageParam],
+        overrides: dict[str, Any] = {},
+    ) -> RetrievalResponse:
+        chat_params = self.get_params(messages, overrides)
+
+        # Generate an optimized keyword search query based on the chat history and the last question
+        query_messages, query_text, filters = await self.generate_search_query(
+            chat_params=chat_params, query_response_token_limit=500
+        )
+
+        # Retrieve relevant items from the database with the GPT optimized query
+        # Generate a contextual and content specific answer using the search results and chat history
```

Comment:
To reduce the repeated comments, you could remove the comments here (and below), since you do have nice docstrings for the functions.